gcc/rust/util/rust-mapping-common.h

```diff
@@ -32,16 +32,31 @@ typedef uint32_t HirId;
 // refers to any top-level decl in HIR
 typedef uint32_t LocalDefId;
 // refers to <Crate><DefId>
-typedef uint64_t DefId;
-
-#define DEF_ID_CRATE_MASK 0xFFFFFFFF00000000
-#define DEF_ID_LOCAL_DEF_MASK 0x00000000FFFFFFFF
+struct DefId
+{
+  CrateNum crateNum;
+  LocalDefId localDefId;
+
+  bool operator== (const DefId &other) const
+  {
+    return this->crateNum == other.crateNum
+	   && this->localDefId == other.localDefId;
+  }
+
+  bool operator!= (const DefId &other) const { return !(*this == other); }
+
+  bool operator< (const DefId &other) const
+  {
+    return ((uint64_t) this->crateNum << 32 | this->localDefId)
+	   < ((uint64_t) other.crateNum << 32 | other.localDefId);
+  }
```

Comment:
I personally do not like operator overloading as I find it hinders code clarity. I'm not a C++ expert so if everyone else agrees with the change, I'm fine with it. I agree that removing the bit shifts and masks is a good thing, but that could also be solved by adding methods

Reply:
I know what you mean @CohenArthur I think we need it for some classes like this one and some others in the name-resolver iirc, since the code at the moment uses a uin64_t and we are using a lot of != and == on it it would be a fairly big refactor I think to change that over.

Reply:
OVerall I think this is ok to use operator overloading at least for now.

Reply:
@CohenArthur and @philberty  Thank you for the feedback. I just realized that I can get rid of the < operator if I use a lambda comparator when declaring the map like map<_, _, decltype(comparator)> in the following line:

https://github.com/Rust-GCC/gccrs/blob/5f0df4812c37fc428b5508e019e9fb7f8a7b77b1/gcc/rust/backend/rust-compile-context.h#L334

I am also attempt getting rid of other operators and do refactor if that is a good approach. I can certainly work on this if it makes sense.

Reply:
Ah ok in that case I would prefer that this is defined in the class or you can define the comparator operators in a separate class instead of another lambda.

Reply:
@philberty and @CohenArthur I have removed operator overloading and added a new class called DefIdComparator. This class has static methods to compare DefId structs. I also refactored code so it does not break.

Not sure if you think the code looks better. I kinda feel the code is a little more complicated now. So I can add operator overloading back if you would like or we can keep the code the way it is.

Once you like the code, I will squash the commits to 1 commit.

Thank you.